src/main/java/org/ibp/api/java/impl/middleware/study/StudyInstanceServiceImpl.java

```diff
@@ -117,66 +116,102 @@ public void deleteStudyInstances(final Integer studyId, final List<Integer> inst
-		final InstanceData instanceData) {
+	public ObservationData updateInstanceObservation(final Integer studyId, final Integer instanceId, final Integer observationDataId,
+		final ObservationData observationData) {
 		this.studyValidator.validate(studyId, true);
 		this.instanceValidator.validateStudyInstance(studyId, Collections.singleton(instanceId));
 
-		final Integer datasetId = this.getEnvironmentDatasetId(studyId);
-		final Integer variableId = instanceData.getVariableId();
+		final Integer datasetId = this.studyService.getEnvironmentDatasetId(studyId);
+		final Integer variableId = observationData.getVariableId();
 		this.datasetValidator.validateExistingDatasetVariables(studyId, datasetId, Collections.singletonList(
 			variableId));
-		this.observationValidator.validateObservationValue(variableId, instanceData.getValue());
-
-		final boolean isEnvironmentCondition =
-			this.datasetService.getDatasetVariablesByType(studyId, datasetId, VariableType.STUDY_CONDITION).stream()
-				.anyMatch(v -> v.getId().equals(variableId));
-		final BindingResult errors = new MapBindingResult(new HashMap<String, String>(), Integer.class.getName());
-		final Optional<InstanceData> existingEnvironmentData =
-			this.middlewareStudyInstanceService.getInstanceData(instanceId, instanceDataId, variableId, isEnvironmentCondition);
-		if (!existingEnvironmentData.isPresent()) {
+		this.observationValidator.validateVariableValue(variableId, observationData.getValue());
+		this.datasetValidator.validateVariableBelongsToVariableType(datasetId, variableId, VariableType.STUDY_CONDITION.getId());
+
+		final BindingResult errors = new MapBindingResult(new HashMap<>(), Integer.class.getName());
+		final Optional<ObservationData> existingObservationData =
+			this.middlewareStudyInstanceService.getInstanceObservation(instanceId, observationDataId, variableId);
+		if (!existingObservationData.isPresent()) {
 			errors.reject(INVALID_ENVIRONMENT_DATA_ID);
-		} else if (!existingEnvironmentData.get().getVariableId().equals(variableId)) {
+		} else if (!existingObservationData.get().getVariableId().equals(variableId)) {
 			errors.reject(INVALID_VARIABLE_FOR_ENVIRONMENT_DATA);
 		}
 
 		if (!errors.getAllErrors().isEmpty()) {
 			throw new ApiRequestValidationException(errors.getAllErrors());
 		}
 
-		instanceData.setInstanceDataId(instanceDataId);
-		instanceData.setInstanceId(instanceId);
-		this.middlewareStudyInstanceService.updateInstanceData(instanceData, isEnvironmentCondition);
-		return instanceData;
+		observationData.setObservationDataId(observationDataId);
+		observationData.setInstanceId(instanceId);
+		this.middlewareStudyInstanceService.updateInstanceObservation(observationData);
+		return observationData;
+	}
+
+	@Override
+	public DescriptorData addInstanceDescriptor(final Integer studyId, final Integer instanceId, final DescriptorData descriptorData) {
+		this.studyValidator.validate(studyId, true);
```

Comment:
The common validations for add/update Instance descriptor data can be consolidated to a method:
`this.studyValidator.validate(studyId, true);
this.instanceValidator.validateStudyInstance(studyId, Collections.singleton(instanceId));

final Integer datasetId = this.studyService.getEnvironmentDatasetId(studyId);
final Integer variableId = descriptorData.getVariableId();
this.datasetValidator.validateExistingDatasetVariables(studyId, datasetId, Collections.singletonList(
			variableId));
this.observationValidator.validateVariableValue(variableId, descriptorData.getValue());
this.datasetValidator.validateVariableBelongsToVariableType(datasetId, variableId, VariableType.ENVIRONMENT_DETAIL.getId());
`